Use transient.el

mathrick opened the pull request:
This updates `magit-p4` to use `transient` instead of `magit-popup`.

Summary of changes:
* Code now uses `transient-define-prefix` instead of `magit-define-popup`, alongside with all the other changes to accommodate the differences between the two libraries
* Keys used have been updated slightly to align them better with keys used by native magit transients for conceptually similar operations
* Bind <kbd>4</kbd> to `magit-p4-popup` in the top-level magit keymap, so it can be invoked without going through `magit-dispatch`. This currently conflicts with the binding for `magit-section-show-level-4`, but I've personally never used that one, unlike `magit-section-show-level-4-all`, so this seems like a worthy sacrifice

Commit:
Initial update to use transient instead of magit-popup

All popups have been rewritten to use transient, but the port is not
fully functional yet and there are known errors remaining.

Commit:
Fully port to transient and recent magit versions

Not all functionality has been fully tested yet, but the basics at
least should be fully functional now.

* Add lexical scoping to magit-p4.el
* Change the sync and submit keystrokes to fp and Pp respectively, to
  mirror magit's own fetch and push
* Ensure the presentation and naming of the transients' infixes is
  consistent with the usual magit naming, rename the variables in the
  source code to match

tarsius commented on magit-p4.el:
```diff
@@ -57,12 +57,12 @@ argument is directory which will hold the Git repository."
            (if (and (not (cl-some (lambda (arg)
                                     (string-match-p "--destination=" arg))
-                                  (magit-p4-clone-arguments)))
+                                  (transient-args (oref transient-current-prefix command))))
```
You should specify the appropriate prefix command explicitly. If a command is a suffix of multiple prefixes, you can also use a list of these prefixes as the value of the PREFIX argument.

It's no accident I haven't made PREFIX optional (and then fallback to `(oref transient-current-prefix command)`; I wanted to make it less appealing to use this "any menu goes" approach.

A user might add a suffix command to some other, to the author unexpected menu. We do not want to use the infix arguments of that unknown menu; they almost certainly are not appropriate.

A user may also invoke the suffix command directly, not from a menu. In that the menu's save, set or default arguments are used. But that is only possible if `transient-args` known which menu it should get the arguments from.

mathrick replied:
Gotcha, fixed.

tarsius commented on magit-p4.el:
```diff
@@ -171,68 +175,91 @@ P4EDITOR and use custom process filter `magit-p4-process-filter'."
+   (lambda (_)
+     (transient-parse-suffixes 'magit-p4-sync-popup magit-p4-sync-clone-shared-arguments))
+   ]
+  ["Actions"
```
It's not necessary to use `:setup-children` to include suffixes as-is. Do this instead:

```elisp
(transient-define-prefix magit-p4-sync-popup ()
  "Pull changes from p4"
  [:description "Options"
   :class transient-column
   magit-p4-sync-clone-shared-options]
  [:description "Arguments"
   :class transient-column
   magit-p4-sync-clone-shared-arguments]
  ["Actions"
   ("p" "Sync" magit-p4-sync)])
```

You could additionally move the group settings to the group definitions:

```elisp
(defvar magit-p4-sync-clone-shared-arguments
  [:description "Options"
   :class transient-column
   ("-b" "Branch" "--branch=")
   ("-c" "Changes files" "--changesfile=" :reader transient-read-existing-file)
   ("-m" "Limit the number of imported changes" "--max-changes=")
   ("-s" "Internal block size to use when iteratively calling p4 changes"
    "--changes-block-size=")
   ("-/" "Exclude depot path" "-/")])

(defvar magit-p4-sync-clone-shared-options
  [:description "Arguments"
   :class transient-column
   ("-d" "Detect branches" "--detect-branches")
   ("-l" "Query p4 for labels" "--detect-labels")
   ("-b" "Import labels" "--import-labels")
   ("-i" "Import into refs/heads/ , not refs/remotes" "--import-local")
   ("-p" "Keep entire BRANCH/DIR/SUBDIR prefix during import" "--keep-path")
   ("-s" "Only sync files that are included in the p4 Client Spec"
    "--use-client-spec")])

(transient-define-prefix magit-p4-sync-popup ()
  "Pull changes from p4"
  [magit-p4-sync-clone-shared-options]
  [magit-p4-sync-clone-shared-arguments]
  ["Actions"
   ("p" "Sync" magit-p4-sync)])
```

mathrick replied:
Thanks, I didn't know variables could just be referenced by name, it was not very obvious from either the docs or the showcase.

I could not get the version with group settings lifted up to the shared options to work, though. It errors out with
```
transient-setup: No applicable method: transient-format, #s(transient-column 1 nil nil nil nil nil nil nil nil ...)
```
I tried with the variables defined both as vectors and lists, but neither works.

tarsius replied:
> I could not get the version with group settings lifted up to the shared options to work, though.

Then stick to the list-of-suffixes approach.

> it was not very obvious from either the docs

Instead of looking into why inline-whole-group approach doesn't work here for you, I'll try to write some more docs. Eventually.

tarsius replied:
Well, I might add some quick notes anyway.

Including a group won't work here:

```elisp
  ["Arguments"
   magit-p4-sync-clone-shared-arguments
   ("-D" "Destination directory" "--destination=" read-directory-name)]
```

Because the variable is followed by a individual binding, the value of the variable must be a list of bindings, not a group. Groups and individual bindings cannot co-exist as siblings. But I got a different error than you got, when I did that anyway. Since you want to insert next to an individual binding anyway, it might be best to just stick to lists of suffixes, and figuring out why using a group didn't work might not be worth the effort.

You might have to wrap the group/list definitions with `eval-and-compile`.

mathrick replied:
Just to make sure I understand what you're saying properly, the code as I have it right now is correct, right? If so, I'm pretty OK with the very minor amount of duplication involved in giving each popup a group for options and arguments separately.

tarsius committed:
Clean up the code, add Makefile

tarsius commented:
(I am switching to "regular posts" because I want to use Forge and that doesn't support "review comments" yet.

> Just to make sure I understand what you're saying properly, the code as I have it right now is correct, right?

No I was trying to say that you should byte-compile to figure out whether you have to use `eval-and-compile`.  I have done that now and the answer is *yes*.

I have started by adding a `Makefile`, which allows me to run the compiler.  (It follows the same basic format as I use for all my packages.  It assumes that all dependencies can be found in git repositories at `../<name>`.  If that doesn't work for you you can adjust the `LOAD_PATH` in `config.mk`.

IMO that's the first step when taking over an existing package: make sure you can build it, so that you can find and address existing issues and to some extend avoid regressions. (Only after that is complete should one move on to improvements, i.e., your switch to `transient` should not have come first.)

So I went on to address existing issues.  Since I mentioned what you should have done differently, I'll also admit that addressing these issues myself was "wrong".  I should have guided you through the process, but that would have taken more time, and I have so many other things to do.

But I left something for you to do: I am not sure my change to `magit-p4-clone` (in 454341c * Enable lexical-binding and fix a bug thus revealed) is correct.  Please check that.

Finally I have rebased your changes on top of my cleanup.

I think it would be better if you squashed your commits, and my fixup commit, into a single commit.  Your original plan to split up the work in "basic changes" and "further improvements" is good, but when it turned out that the first of these had issues (the `transient-args` thing), that shouldn't be addressed in a new commit.  The original commit should be fixed instead.  Maybe your plan was to do that later and just added a new commit to make it easier for me to review.  Other maintainers will appreciate that, but I prefer when commits are fixed immediately. I've you wish you can keep the two original commits, but that's more work because you'll have to resolve more conflicts during rebase, so I won't insist on that.

tarsius commented:
Please open a pr in their repository to address these:

```elisp
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘case’ is an obsolete alias (as of 27.1); use ‘cl-case’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘pushnew’ is an obsolete alias (as of 27.1); use ‘cl-pushnew’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘case’ is an obsolete alias (as of 27.1); use ‘cl-case’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘case’ is an obsolete alias (as of 27.1); use ‘cl-case’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘incf’ is an obsolete alias (as of 27.1); use ‘cl-incf’ instead.
../p4/p4.el: Warning: ‘incf’ is an obsolete alias (as of 27.1); use ‘cl-incf’ instead.
../p4/p4.el: Warning: ‘incf’ is an obsolete alias (as of 27.1); use ‘cl-incf’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘incf’ is an obsolete alias (as of 27.1); use ‘cl-incf’ instead.
../p4/p4.el: Warning: ‘incf’ is an obsolete alias (as of 27.1); use ‘cl-incf’ instead.
../p4/p4.el: Warning: ‘incf’ is an obsolete alias (as of 27.1); use ‘cl-incf’ instead.
../p4/p4.el: Warning: ‘case’ is an obsolete alias (as of 27.1); use ‘cl-case’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
../p4/p4.el: Warning: ‘loop’ is an obsolete alias (as of 27.1); use ‘cl-loop’ instead.
```

mathrick commented:
Sorry, I've been pretty busy since last week. I'm expecting to have some time to look at and properly understand your comments properly this weekend.

mathrick commented on magit-p4.el:
```diff
@@ -1,10 +1,10 @@
 ;; Author: Damian T. Dobroczyński <qoocku@gmail.com>
 ;; Maintainer: Aleksey Fedotov <lexa@cfotr.com>
-;; Package-Requires: ((emacs "25.1") (magit "2.1") (magit-popup "2.1") (p4 "12.0") (cl-lib "0.5"))
+;; Package-Requires: ((emacs "25.1") (magit "4.3.0") (transient "0.8.4") (p4 "12.0") (cl-lib "0.5"))
```
Is there a reason you've set the requirements to the absolutely latest and greatest version for magit and transient? I don't like aggressively recent requirements without a compelling argument, I find it unkind to the users to say that they _must_ have the version that just got released if the code would work just fine with a much older requirement.

The versions I originally picked (`3.3.0` and `0.4.3`) roughly correspond to the oldest versions I have lying around on any of my machines and which I could test; is there anything wrong with those? If the code remains compatible with the latest versions while targetting old ones, I'd prefer to keep the oldest versions possible as the requirement.

tarsius replied:
This is a major change and to me it makes sense to take the opportunity to make a clear cut.

Supporting very old versions has a very real cost, when done properly (i.e., not just claiming that old versions are supported when in fact it is unknown whether they are), and there are other areas where a maintainer's time is better invested. Stopping to support some old version is more difficult than to not support some old version from the get go. This is an existing package, but its a major upgrade so it is as close to "from the get to" as we can get here.

I do not know what the actual minimally required versions are. Figuring that out would be a massive time investment. I think it is unkind to the user to randomly pick some random version (e.g., "the oldest versions I have lying around"). They might end up sticking to what they have installed, assuming that the specified minimal versions are actually accurate, and then run into issues because the specified required versions were made up.

By requiring users to use a recent version, you force them to make the minimal time investment necessary to update. As an additional benefit they get a ton of fixed bugs (and sure, likely also some regressions).

Magit and Transient have many users, and supporting them takes a lot of time. I am barely able to keep up. So when someone reports a bug, I expect them to help me help them, and that involves updating to the latest releases. You would be surprised how often users report issues that have been fixed long ago. By "forcing" users to use a recent version up front, I save them from getting an initial response of "please update" when they report an issue, and then having to wait for me to get around to look at the report a second time.

Declaring that some older versions are supported, only helps users if the maintainer actually tests that this is still true (regularly, ideally automated). For my major packages I do that when it comes to the Emacs dependency. The cost of supporting old Emacs releases is large. It prevents the use of newer Emacs features and forces the use of workarounds for Emacs bugs that have been fixed long ago. I am not doing that for dependencies that are individual package, not even (or especially) my own packages.

Magit has many users but only one maintainer. I am just human not RHEL. If someone wants RHEL-like stability, they can pay my accordingly.

Magit 3.3.0 is 3.5 years old. It is not reasonable to still be using that. Users who get Emacs package with their distributions package manager, should question whether that is a reasonable thing to do, if it turns out that their distribution is stuck on such an old version.

Transient 0.4.3 has very serious issues. "Emacs accepts no input and has to be killed from the outside" kinda serious.

A reasonable compromise would be to require 4.0.0 and 0.8.0. If a user used these versions and reported an issue, I would still first ask them to update, but at least it is quite possibly safe to assume, that this package works with these versions.

You also have to bump the Emacs dependency. Magit 4.0.0 requires 26.1. Magit >= 4.2.0 requires 27.1. For that reason it is actually a good idea to only require 4.0.0, not 4.3.0, here for a while.

mathrick replied:
Apologies for the delay, I've been very busy and split between projects, and just wasn't able to get through all the items I wanted to get done on this PR. That's why it took me until now for me to have a useful response. I do expect to remain busy for the forseeable future, but I should be able to have every other weekend to work on this package as needed. 

> Declaring that some older versions are supported, only helps users if the maintainer actually tests that this is still true (regularly, ideally automated). For my major packages I do that when it comes to the Emacs dependency. The cost of supporting old Emacs releases is large. It prevents the use of newer Emacs features and forces the use of workarounds for Emacs bugs that have been fixed long ago. I am not doing that for dependencies that are individual package, not even (or especially) my own packages.

Granted. If it's not tested automatically, it should be assumed broken. I added some basic tests and a github workflow to test it on a matrix of OS x Emacs versions. I also moved to Eldev as the build/test manager, since it's more integrated and also supports non-Linux platforms.

The tests don't do much at the moment, but they're already enough to get GH actions to do what they do best, ie. trigger mysterious and non-repeatable failures, so I'm currently having fun debugging those. Once I have a workflow that repeatably passes, I'll expand it to provide a meaningful coverage of functionality under different Emacs and dependency versions.

> Transient 0.4.3 has very serious issues. "Emacs accepts no input and has to be killed from the outside" kinda serious.
> 
> A reasonable compromise would be to require 4.0.0 and 0.8.0. If a user used these versions and reported an issue, I would still first ask them to update, but at least it is quite possibly safe to assume, that this package works with these versions.
> 
> You also have to bump the Emacs dependency. Magit 4.0.0 requires 26.1. Magit >= 4.2.0 requires 27.1. For that reason it is actually a good idea to only require 4.0.0, not 4.3.0, here for a while.

Thank you, that is a very solid reason and a good compromise. I am perfectly OK with Emacs 26.x or 27.x being the minimum; 27.2 is the version shipped with Ubuntu 20.04, which is still supported (even if just barely), so I think that's fine floor to set, though I'll try to test it on 26.3 if I can get the workflow to function.

I do agree that there is such a thing as too much maintenance effort, and if something is just infeasible to support on an old/buggy version, then it might be necessary to require a newer one. But I also hate spurious requirements that force users to go through the hoops of figuring out how to get the latest and greatest on their machine. Especially if it's a work laptop with arbitrary IT requirements, which is very much not out of question for Perforce users. I've got a work laptop with the version of Ubuntu that IT installed :)

tarsius replied:
> Especially if it's a work laptop with arbitrary IT requirements, which is very much not out of question for Perforce users.

Good point.

mathrick reviewed (commented):
I will squash the commits before merging this in, but I'm leaving them as-is for now because it makes it easier to ask questions about your changes.

mathrick committed:
Use Eldev for testing and release, add test skeleton

- Eldev provides integrated build / test / release cycle management
  with support for CI workflows
- Buttercup is used as the testing framework; Eldev has built-in
  integration for running and reporting Buttercup tests
- Rudimentary test suite currently only really checks that the package
  can be loaded and adds a key binding in the Magit dispatch transient

mathrick committed:
Add Github Actions workflow

mathrick committed:
Add support for testing compatibility with old dependency versions

mathrick committed:
Drop compatibility with Emacs 26.x

Magit 4.0.0 requires 27.1, and even Ubuntu 20.04 ships with 27.2.
Plus 26.3 is incompatible with multi-line package dependency declarations

mathrick commented:
@tarsius: I've finished making the changes necessary to automatically test for backwards compatibility and I now have a passing CI workflow run on [every version going back to 27.2](https://github.com/mathrick/magit-p4/actions/runs/13879896836), so I consider the PR to be ready to merge. I still want to add a full, proper test suite, rather than just a skeleton, but I think it's reasonable not to block on that, given that the code is already in a much better shape when it comes to being testable and tested than it was when I took over. I dropped 26.3 from compatibility list, since Magit 4.0.0 requires 27.1 anyway, and 26.x is truly ancient now.

Also, I didn't get your invite before it expired, since the notification went to spam for some reason; would you mind resending it?

mathrick commented:
@tarsius: any objections to me merging this in?

tarsius commented on tests/helpers.el:
```diff
@@ -0,0 +1,35 @@
+                                      (listify-key-sequence (kbd keys)))))
+
+
+(provide 'helpers)
```
This is a test utility, not a library intended for end-users and expected to follow conventions such as namespaced symbol prefixes. So it shouldn't provide a feature, and it should be loaded using `load`. Maybe you can use a relative path or maybe you will have to involve `(or load-file-name buffer-file-name)` in the path construction.

While that isn't strictly necessary, as long as you don't provide a feature, I think the file-name should be namespaced. "helper" is a reasonable name, but anyone else may chose the same name.

mathrick replied:
Is this really a concern though? The file is already located under `tests/`, so it's not going to be in the `load-path` in the first place, will it?

tarsius replied:
The tool I use extract metadata for the Emacsmirror would find it and add `helpers` to the features provided by this package. Another way to combat that is to add `tests/.nosearch`. It does this because some package repositories have libraries in more than one location (e.g., the top-level as well as `extensions/`).

mathrick replied:
Either way, done.

mathrick replied:
Ah, that makes sense, I didn't think of other code scanning it.

tarsius commented on package-lockfile.el:
```diff
@@ -0,0 +1,7 @@
+ ("with-editor" . "v3.4.1")
+ ;; Transitive
+ ("dash" . "2.19.1"))
+:gamma
```
Melpa would include this in the package, so you would have to edit the Melpa recipe to exclude it again. Can the value be provided directly in `straight-profiles` or does that insist on a file? Is this only used during testing, and if so, might it not make more sense to place it in `tests/` or `.github/workflows/`? If located there, the recipe would not have to be adjusted. If you keep it at the top-level, consider making it a hidden file (the recipe would have to be adjusted anyway [though that's something I might want to change in Melpa/`package-build.el`, after careful consideration]).

mathrick replied:
It's test-only, I'll move it there.
